### ISSUE TITLE
NET-1788 New Rule S1909: "goto" statements should not be used to jump into blocks

### ISSUE THREAD

**Issue by gregory-paidis-sonarsource** [SonarSource/sonar-dotnet#6661]
S1909

[CSharp spec for labels](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#125-labeled-statements)
[CSharp spec for goto](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#12104-the-goto-statement)

From the spec:
> Note: This means that goto statements can transfer control within blocks and out of blocks, but never into blocks.

**Comment by martin-strecker-sonarsource** [comment]
I think the issue is only valid for VB. The example given in the c-family section does not compile in C# due to the scoping rules.
```c
void f1 (int a) {
  if (a <=0) {
    goto L2;  // Noncompliant; jumps into a different block
  }
  for (int i = 0; i < a; i++) {
    L2:;
    //...  Should only have come here with a >=0. Loop is infinite if a < 0
  }
}
```

The issue says "`goto` statements should not be used to jump into blocks", but the CSharp spec does not allow jumps into blocks and raises `error CS0159: No such label 'xxx' within the scope of the goto statement`.

> Note: This means that goto statements can transfer control within blocks and out of blocks, but never into blocks.
[Source](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#125-labeled-statements)

[Example on sharplab](
https://sharplab.io/#v2:CYLg1APgAgzABFATHAwnA3gWAFBzwgFjgFkAKASwDsAXOAQwEod8NmX9yAzOUuuAHgC8ABia52eLOIksA5gHtq8uABkAjAG48Aem1wAcvMoBjeQFsADgBtydGloBWAV0sBnOFSX04wLpwCmAE7+NHAARlbyxgDWbBIAvnHsnPKBPJ4ecIJwwlrkAvR5YGBiMqzSZapqIBpJLIkVDfFAA)

For VB it also fails in some cases:
[Example for VB on sharplab](https://sharplab.io/#v2:DYLgbgRgNAJiDUAfACgVwsAlgYwAQGFgBDAZxIIFgAoXW3ZAJ0zCIBcBTXAZXVwFkAFACEAngDUiwXEVwBBcgEkAdhwDm7BgEpqdXbgUAzabgA8AXlwAGXABUAFuyU69LgOIB7G+9wAZAIzOLgCiSjD6BtSBegBi7gy4mHKKKuzq8RbWXsYAtLgBVP4gUboAcuwAHqxRIWE8ENQ1BMRk1EA=)

But in others, it compiles:
[sharplab](https://sharplab.io/#v2:DYLgbgRgNAJiDUAfACgVwsAlgYwAQGFgBDAZxIIFgAoXW3ZAJ0zCIBcBTXAZXVwFkAFEVwBBcgEkAdhwDm7BlFwRcpXFNnyAlNTq61AMxW4APAF5cABlwAVABbtJOvc4DiAe2tvcAGQCMT5wBRSRgDagC9cUNlAD5LG3tHKj8QCN0HUMx9CODQnghqXIJiMmogA=)

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1788](https://sonarsource.atlassian.net/browse/NET-1788)

[NET-1788]: https://sonarsource.atlassian.net/browse/NET-1788?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by pavel-mikula-sonarsource** [comment]
Moved
